hs_core/tasks.py

```diff
@@ -405,6 +405,74 @@ def copy_resource_task(ori_res_id, new_res_id=None, request_username=None):
+        new resource will be created.
+        username: the requesting user's username
+    Returns:
+        the new versioned resource url as the payload
+    """
+    try:
+        new_res = None
+        if not new_res_id:
+            new_res = create_empty_resource(ori_res_id, username)
+            new_res_id = new_res.short_id
+        utils.copy_resource_files_and_AVUs(ori_res_id, new_res_id)
+
+        # copy metadata from source resource to target new-versioned resource except three elements
+        ori_res = utils.get_resource_by_shortkey(ori_res_id)
+        if not new_res:
+            new_res = utils.get_resource_by_shortkey(new_res_id)
+        utils.copy_and_create_metadata(ori_res, new_res)
+
+        # add or update Relation element to link source and target resources
+        hs_identifier = new_res.metadata.identifiers.all().filter(name="hydroShareIdentifier")[0]
+        ori_res.metadata.create_element('relation', type='isReplacedBy', value=hs_identifier.url)
+
+        if new_res.metadata.relations.all().filter(type='isVersionOf').exists():
+            # the original resource is already a versioned resource, and its isVersionOf relation
+            # element is copied over to this new version resource, needs to delete this element so
+            # it can be created to link to its original resource correctly
+            eid = new_res.metadata.relations.all().filter(type='isVersionOf').first().id
+            new_res.metadata.delete_element('relation', eid)
+
+        hs_identifier = ori_res.metadata.identifiers.all().filter(name="hydroShareIdentifier")[0]
+        new_res.metadata.create_element('relation', type='isVersionOf', value=hs_identifier.url)
+
+        if ori_res.resource_type.lower() == "collectionresource":
+            # clone contained_res list of original collection and add to new collection
+            # note that new version collection will not contain "deleted resources"
+            new_res.resources = ori_res.resources.all()
+
+        # create bag for the new resource
+        create_bag(new_res)
+
+        # since an isReplaceBy relation element is added to original resource, needs to call
+        # resource_modified() for original resource
+        utils.resource_modified(ori_res, by_user=username, overwrite_bag=False)
+        # if everything goes well up to this point, set original resource to be immutable so that
+        # obsoleted resources cannot be modified from REST API
+        ori_res.raccess.immutable = True
+        ori_res.raccess.save()
+        ori_res.save()
+        # release the lock if new version of the resource is created successfully
+        ori_res.locked_time = None
+        ori_res.save()
+        return new_res.get_absolute_url()
+    except Exception as ex:
+        if new_res:
+            new_res.delete()
+        # release the lock if new version of the resource is created successfully
```

Comment:
This comment is incorrect and looks like it was just copy/pasted from above.  I'm weighing whether the `locked_time=None` with the save should be put in a finally block.  The only scenario I see where it would be needed is if an exception is thrown when deleting a resource... which could happen.  If that does happen `ori_res` would be locked until some manual intervention.  I think it needs to be in a finally block.

Reply:
Good catch on the wrong copy and pasted comment and using finally to capture the potential edge cases. Will update code accordingly.